snow/networking/worker/pool.go

```diff
@@ -12,51 +12,72 @@ type Request func()
 type Pool interface {
 	// Send the request to the worker pool.
 	//
-	// Send should never be called after [Shutdown] is called.
+	// Send can be safely called after [Shutdown] and it'll be no-op.
 	Send(Request)
 
 	// Shutdown the worker pool.
 	//
 	// This method will block until all workers have finished their current
 	// tasks.
 	//
-	// It is safe to call shutdown multiple times.
+	// Shutdown can be safely called multiple times.
 	Shutdown()
 }
 
 type pool struct {
 	requests chan Request
 
+	// [shutdownOnce] ensures Shutdown idempotency
 	shutdownOnce sync.Once
-	shutdownWG   sync.WaitGroup
+
+	// [shutdownWG] makes sure all workers have stopped before Shutdown returns
+	shutdownWG sync.WaitGroup
+
+	// closing [quit] tells the workers to stop working
+	quit chan struct{}
 }
 
-func NewPool(size int) Pool {
+func NewPool(workersCount int) Pool {
 	p := &pool{
 		requests: make(chan Request),
+		quit:     make(chan struct{}),
 	}
-	p.shutdownWG.Add(size)
-	for w := 0; w < size; w++ {
+
+	for w := 0; w < workersCount; w++ {
+		p.shutdownWG.Add(1)
 		go p.runWorker()
 	}
+
 	return p
 }
 
 func (p *pool) runWorker() {
 	defer p.shutdownWG.Done()
 
-	for request := range p.requests {
-		request()
+	for {
+		select {
+		case <-p.quit:
+			return // stop worker
+		case request := <-p.requests:
+			if request != nil {
+				request()
+			}
+		}
+	}
+}
+
+func (p *pool) Send(msg Request) {
+	select {
+	case <-p.quit:
+	case p.requests <- msg:
 	}
 }
 
 func (p *pool) Shutdown() {
 	p.shutdownOnce.Do(func() {
-		close(p.requests)
+		close(p.quit)
+		// We don't close requests channel to avoid panics
+		// upon sending request over a closed channel.
```

Comment:
nit: move this comment above the `close` line. I personally would probably not comment this because i feel like this line is self-documenting but leaving this inclusion of this to your preference

Reply:
We also no longer close `p.requests` here, is that okay?

Reply:
if we close p.request we risk a panic if another goroutine invoke Send. I tried to express this in the comment